### PR TITLE
[mutable-arrays] allow state effects in jit by building in run_state

### PR DESCRIPTION
**Pull request by mattjj**
with help from @sharadmv, @yashk2810, @dougalm, and others

This is an experiment based on @dougalm's `MutableArray` proposal in go/jax-oop-proposal (similar to something we've considered adding for a long time). See [the test](https://github.com/google/jax/pull/20026/files#diff-345fab73b3320212e0556e11f83a3cb1be42ece90f54c04633383d23e34452bdR1542-R1553).

The basic strategy is to apply discharge_state when lowering a jaxpr with state effects to HLO, and update the dispatch path accordingly. Specifically:
1. in tests only for now, introduce a MutableArray data type;
2. teach jit to abstract it to a Ref(ShapedArray) type, register an input handler, etc;
3. call discharge_state in `lower_sharding_computation` to lower a jaxpr with refs to a jaxpr (and then to an HLO) with extra outputs, and set up aliasing;
4. teach the output side of the dispatch path to drop those outputs.

As an alternative to (3), we could potentially lower away the effects at a higher level, like in _pjit_lower_cached. They are similar because _pjit_lower_cached is the only (non-xmap) caller of lower_sharding_computation. I decided to do it in lower_sharding_computation mainly because that's closer to where we set up aliases, and I wanted to make mutable arrays correspond to aliased inputs/outputs on the XLA computation.

### CODE REVIEW

**Comment by yashk2810** [comment]
> lower away the effects at a higher level, like in _pjit_lower_cached

Yes, once xmap goes away, I plan to clean this up and lift it to pjit. Right now, DCE and now this discharge happens in lower_sharding_computation. So it's good to keep all of this in 1 place and cleanup later.

File: jax/_src/core.py [hunk 1]
```diff
@@ -203,6 +203,7 @@ def replace(self, *, constvars=None, invars=None, outvars=None, eqns=None,
     outvars = self.outvars if outvars is None else outvars
     eqns = self.eqns if eqns is None else eqns
     effects = self.effects if effects is None else effects
+    debug_info = self.debug_info if debug_info is None else debug_info
```
Comment (mattjj) [inline jax/_src/core.py:206]: This was an unrelated change that I happened to notice.

**Review by yashk2810 — APPROVED** [review]

**Review by sharadmv — APPROVED** [review]
top level
